### PR TITLE
Premium bug fix

### PR DESCRIPTION
**Pull request by HitoriWoB**
Fixes:
Chat name color bug on premium page
Allows user admins to select red again

Patches:
Infinite level 1 forbidden seal

### CODE REVIEW

File: chat.php [hunk 1]
```diff
@@ -49,15 +49,53 @@ function chat() {
...
 			$supported_colors = array(
-				'blue' => -1,
+				'black' => 0,'blue' => -1,
 				'pink' => -2,
 				'gold' => -3,
 				'green' => -4,
 				'teal' => -5,
 				'purple' => -6,
 				'red' => -7,
 			);
```
Comment (levimeahan) [inline chat.php:59]: While we're here, can we get rid of this string => number mapping and just store the color names in the DB?

File: chat.php [hunk 2]
```diff
@@ -49,15 +49,53 @@ function chat() {
...
+                case 'red':
+                    if(!$player->isUserAdmin() || !$player->isHeadAdmin()) {
+                        $player->chat_color = 'black';
+                    }
+                    break;
+                default:
+                    break;
+            }
+
+			$user_color = (!$player->canChangeChatColor()) ? 0 : ($supported_colors[$player->chat_color]);
```
Comment (levimeahan) [inline chat.php:98]: per above
```suggestion
			$user_color = (!$player->canChangeChatColor()) ? 0 : ($player->chat_color);
```

File: chat.php [hunk 3]
```diff
@@ -49,15 +49,53 @@ function chat() {
...
+                    }
+                    break;
+                case 'red':
+                    if(!$player->isUserAdmin() || !$player->isHeadAdmin()) {
+                        $player->chat_color = 'black';
+                    }
+                    break;
+                default:
+                    break;
+            }
```
Comment (levimeahan) [inline chat.php:96]: Hmm what if we centralize the name colors as defined in premium.php to `User` and then just do an `if(!in_array($player->chat_color, $player->chat_color_options))` check?

https://github.com/elementum-games/shinobi-chronicles/blob/9874185125d13e20f86730561728812046576a11/premium.php#L974-L1029

File: premium.php [hunk 4]
```diff
@@ -528,53 +528,53 @@ function premium() {
 		}
 		$system->printMessage();
 	}
-	else if(isset($_POST['change_color']) && ($player->forbidden_seal || $player->premium_credits_purchased)) {
+	else if(isset($_POST['change_color']) && $player->canChangeChatColor()) {
 		$color = $system->clean($_POST['name_color']);
 		switch($color) {
```
Comment (levimeahan) [inline premium.php:533]: same comment as above - if we move color options to an array defined in `User`, then we can do a single array check here

**Comment by HitoriWoB** [comment]
Changes made

File: classes/User.php [hunk 5]
```diff
@@ -1254,6 +1273,66 @@ public function getAvatarSize(): int {
...
+        }
+
+        return false;
+    }
+
+    /**
+     * @param false $forceAll // should only be used for debugging
+     * @return string[]
+     */
+    public function getNameColors($forceAll = false) {
```
Comment (levimeahan) [inline classes/User.php:1299]: Head admin should be able to use all colors, is this debug code worth pushing to production?

**Pull request closed without merging**